sdks/go/pkg/beam/runners/dataflow/dataflowlib/metrics.go

```diff
@@ -0,0 +1,125 @@
+// Monitoring. Gauge metrics are not supported. The output metrics.Results will
+// not contain any gauges.
+func FromMetricUpdates(allMetrics []*df.MetricUpdate, job *df.Job) *metrics.Results {
+	ac, ad := groupByType(allMetrics, job, true)
+	cc, cd := groupByType(allMetrics, job, false)
+
+	return metrics.NewResults(metrics.MergeCounters(ac, cc), metrics.MergeDistributions(ad, cd), make([]metrics.GaugeResult, 0))
+}
+
+func groupByType(allMetrics []*df.MetricUpdate, job *df.Job, tentative bool) (
+	map[metrics.StepKey]int64,
+	map[metrics.StepKey]metrics.DistributionValue) {
+	counters := make(map[metrics.StepKey]int64)
+	distributions := make(map[metrics.StepKey]metrics.DistributionValue)
+
+	for _, metric := range allMetrics {
+		isTentative := metric.Name.Context["tentative"] == "true"
+		// Returns true when variables differ (exclusive or)
+		if (isTentative || tentative) && (!isTentative || !tentative) {
+			continue
+		}
+
+		key, err := extractKey(metric, job)
+		if err != nil {
+			continue
+		}
+
+		if metric.Scalar != nil {
+			v, err := extractCounterValue(metric.Scalar)
+			if err != nil {
+				continue
+			}
+			counters[key] = v
+		} else if metric.Distribution != nil {
+			v, err := extractDistributionValue(metric.Distribution)
+			if err != nil {
+				continue
+			}
+			distributions[key] = v
+		}
+	}
+	return counters, distributions
+}
+
+func extractKey(metric *df.MetricUpdate, job *df.Job) (metrics.StepKey, error) {
+	stepName, ok := metric.Name.Context["step"]
+	if !ok {
+		return metrics.StepKey{}, fmt.Errorf("could not find the internal step name")
+	}
+	userStepName := ""
+
+	for _, step := range job.Steps {
+		if step.Name == stepName {
+			properties := make(map[string]string)
+			json.Unmarshal(step.Properties, &properties)
+			userStepName = properties["user_name"]
+			break
+		}
+	}
+	if userStepName == "" {
+		return metrics.StepKey{}, fmt.Errorf("could not translate the internal step name %v", stepName)
+	}
+
+	namespace := metric.Name.Context["namespace"]
+	if namespace == "" {
+		namespace = "dataflow/v1b3"
+	}
+
+	return metrics.StepKey{Step: userStepName, Name: metric.Name.Name, Namespace: namespace}, nil
+}
+
+func extractCounterValue(obj interface{}) (int64, error) {
+	v, ok := obj.(float64)
+	if !ok {
+		return -1, fmt.Errorf("expected float64, got data of type %T instead", obj)
+	}
+	return int64(v), nil
+}
+
+func extractDistributionValue(obj interface{}) (metrics.DistributionValue, error) {
+	m := obj.(map[string]interface{})
+	propertiesToVisit := []string{"count", "sum", "min", "max"}
+	values := make([]int64, 4)
+
+	for i, p := range propertiesToVisit {
+		v, ok := m[p].(float64)
+		if !ok {
+			return metrics.DistributionValue{}, fmt.Errorf("expected float64, got data of type %T instead", m[p])
+		}
+		values[i] = int64(v)
+	}
+	return metrics.DistributionValue{Count: values[0], Sum: values[1], Min: values[2], Max: values[3]}, nil
+}
```

Comment:
I love this implementation. That is all.

Reply:
Good to hear that!